File: src/main/java/sujung/linkedlist/RemoveDuplicatesFromSortedList.java
package sujung.linkedlist;

public class RemoveDuplicatesFromSortedList {

    private static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }


    public ListNode deleteDuplicates(ListNode head) {
        ListNode tempHead = head;
        while (tempHead != null) {
            while (tempHead.next != null && tempHead.val == tempHead.next.val) {
                tempHead.next = tempHead.next.next;
            }
            tempHead = tempHead.next;
        }
        return head;
    }
}

File: src/main/java/thkim/linkedlist/RemoveDuplicatesFromSortedList.kt
package thkim.linkedlist

fun deleteDuplicates(head: ListNode?): ListNode? {
    if (head == null) {
        return head
    }

    val answer = ListNode(head.`val`)
    var nextNode = head.next

    fun recursion(curNode: ListNode) {
        if (nextNode != null) {
            if (curNode.`val` == nextNode!!.`val`) {
                nextNode = nextNode!!.next

                recursion(curNode)
            } else {
                curNode.next = nextNode

                recursion(curNode.next!!)
            }
        } else {
            curNode.next = null
        }
    }
    recursion(answer)

    return answer
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}